File: 알고리즘_도감/sgoldenbird/week3.js
// 1. 정렬의 기본: 최소값 찾기 
function solution1(arr) {
  let min = arr[0];
  for (let i = 1; i < arr.length; i++) {
    if (arr[i] < min) {
        min = arr[i];
    }
  }

  return min;
}

// 2. 버블 정렬: 인접 요소 교환
function solution2(arr) {
  for (let i = 0; i < arr.length-1; i++) {

    for (let j = 0; j< arr.length-1-i; j++ ){
      if (arr[j] > arr[j + 1]) {
        let temp = arr[j];
        arr[j] = arr[j + 1];
        arr[j + 1] = temp;
      }
    }
  }

  return arr[arr.length - 1];
}

// 3. 선택 정렬: 최소값 선택
function solution3(arr) {
  for (let i = 0; i < arr.length-1; i++) {
        let minIndex = i;

    for (let j = i + 1; j < arr.length; j++) {
      if (arr[j] < arr[minIndex]) {
          minIndex = j;
        }
      }
  
    if (minIndex !== i) {
      let temp = arr[i];
      arr[i] = arr[minIndex];
      arr[minIndex] = temp;
    }
  }

  return arr[1];
}


// 4. 삽입 정렬: 삽입 위치 찾기
function solution4(arr) {
  for (let i = 1; i < arr.length; i++) {
        let currentVal = arr[i]; 
        let j = i - 1; 

        while (j >= 0 && arr[j] > currentVal) {
            arr[j + 1] = arr[j]; 
            j--; 
        }

        arr[j + 1] = currentVal;
    }

  return arr[0];
}


// 5. 힙 정렬: 최대값 추출
function swap(arr, i, j) {
  let temp = arr[i];
  arr[i] = arr[j];
  arr[j] = temp;
}

function maxHeapify(arr, size, i) {
  let largest = i;       
  let left = 2 * i + 1;  
  let right = 2 * i + 2; 

  if (left < size && arr[left] > arr[largest]) {
      largest = left;
  }

  // 오른쪽 자식이 존재하고, 현재까지 가장 큰 값보다 크다면 largest 갱신
  if (right < size && arr[right] > arr[largest]) {
      largest = right;
  }

  // 부모보다 큰 자식이 있다면 위치를 교환하고, 자식 노드로 내려가며 재귀적으로 heapify 수행
  if (largest !== i) {
      swap(arr, i, largest);
      maxHeapify(arr, size, largest);
  }
}

function solution5(arr) {
  let heapSize = arr.length;

  // 1. 최대 힙 생성 (Build Heap)
  // 자식이 있는 마지막 부모 노드(heapSize / 2 - 1)부터 역순으로 heapify를 수행합니다.
  for (let i = Math.floor(heapSize / 2) - 1; i >= 0; i--) {
      maxHeapify(arr, heapSize, i);
  }

  // 2. 가장 큰 값(루트 노드, 인덱스 0)을 배열의 맨 뒤로 보냅니다.
  swap(arr, 0, heapSize - 1);
  
  // 힙 크기를 1 줄여서 가장 큰 값을 힙에서 제외합니다.
  heapSize--;

  // 3. 루트 노드가 바뀌었으므로 다시 힙 속성을 만족하도록 재정렬합니다.
  maxHeapify(arr, heapSize, 0);

  // 4. 이제 다시 루트 노드(인덱스 0)에 두 번째로 큰 값이 위치하게 됩니다.
  return arr[0];
}

// 6. 병합 정렬: 분할 정복

// 배열을 반으로 쪼개는 재귀 함수
function mergeSort(arr) {
  if (arr.length <= 1) {
      return arr;
  }

  // 중간 지점을 찾아 배열을 왼쪽과 오른쪽으로 나눕니다.
  const mid = Math.floor(arr.length / 2);
  const left = arr.slice(0, mid);
  const right = arr.slice(mid);

  // 쪼갠 양쪽을 각각 다시 정렬한 후 병합합니다.
  return merge(mergeSort(left), mergeSort(right));
}

// 정렬된 두 배열을 하나로 합치는 함수
function merge(left, right) {
  const result = [];
  let i = 0; // left 배열을 가리키는 포인터
  let j = 0; // right 배열을 가리키는 포인터

  while (i < left.length && j < right.length) {
      if (left[i] <= right[j]) {
          result.push(left[i]);
          i++;
      } else {
          result.push(right[j]);
          j++;
      }
  }

  // 한쪽 배열이 먼저 비었을 때, 남은 요소들을 결과 배열에 뒤이어 붙여줍니다.
  return result.concat(left.slice(i)).concat(right.slice(j));
}

function solution6(arr) {
  // 1. 병합 정렬을 통해 배열을 오름차순으로 정렬합니다.
  const sortedArr = mergeSort(arr);
  
  let midIndex;
  
  // 2. 중간 인덱스를 계산합니다.
  if (sortedArr.length % 2 === 0) {
      // 길이가 짝수일 때 (예: 길이 4 -> 인덱스 1, 2 중 왼쪽인 1 선택)
      midIndex = ( sortedArr.length / 2) - 1;
  } else {
      midIndex = Math.floor( sortedArr.length / 2);
  }
  
  return sortedArr[midIndex];
}

// 7. 퀵 정렬: 피벗 선택
function quickSort(arr) {
    if (arr.length <= 1) {
        return arr;
    }

    const pivot = arr[0]; // 문제 조건에 따라 첫 번째 요소를 피벗으로 선택합니다.
    const left = [];
    const right = [];

    for (let i = 1; i < arr.length; i++) {
        if (arr[i] < pivot) {
            left.push(arr[i]);  // 피벗보다 작은 값은 왼쪽 배열로
        } else {
            right.push(arr[i]); // 피벗보다 크거나 같은 값은 오른쪽 배열로
        }
    }

    // 쪼개진 왼쪽과 오른쪽 배열을 각각 다시 퀵 정렬하고, 피벗과 하나로 합칩니다.
    return [...quickSort(left), pivot, ...quickSort(right)];
}

function solution7(arr) {
    const sortedArr = quickSort(arr);

    const firstElement = sortedArr[0];
    const lastElement = sortedArr[sortedArr.length - 1];

    return firstElement + lastElement;
}

File: 알고리즘_도감/gnsals0904/week3.java
class week3 {
    public static void main(String[] args) {
        System.out.println(solution1(new int[] {3, 1, 4, 1, 5, 9, 2, 6}));
        System.out.println(solution2(new int[] {64, 34, 25, 12, 22, 11, 90}));
        System.out.println(solution3(new int[] {17, 5, 12, 8, 1}));
        System.out.println(solution4(new int[] {5, 2, 4, 6, 1, 3}));
        System.out.println(solution5(new int[] {4, 10, 3, 5, 1}));
        System.out.println(solution6(new int[] {38, 27, 43, 3, 9, 82, 10}));
        System.out.println(solution7(new int[] {10, 7, 8, 9, 1, 5}));
    }

    /**
     * 1. 정렬의 기본: 최소값 찾기
     *   - 배열을 순회하며 현재까지 가장 작은 값을 갱신
     * @param arr
     * @return
     */
    static int solution1(int[] arr) {
        int min = arr[0];

        for (int i = 1; i < arr.length; i++) {
            if (arr[i] < min) {
                min = arr[i];
            }
        }

        return min;
    }

    /**
     * 2. 버블 정렬: 인접 요소 교환
     *   - 인접한 두 값을 비교해서 큰 값을 오른쪽으로 보냄
     *   - 정렬 후 마지막 요소를 반환
     * @param arr
     * @return
     */
    static int solution2(int[] arr) {
        bubbleSort(arr);
        return arr[arr.length - 1];
    }

    static void bubbleSort(int[] arr) {
        for (int i = 0; i < arr.length - 1; i++) {
            for (int j = 0; j < arr.length - 1 - i; j++) {
                if (arr[j] > arr[j + 1]) {
                    swap(arr, j, j + 1);
                }
            }
        }
    }

    /**
     * 3. 선택 정렬: 최소값 선택
     *   - 남은 구간에서 가장 작은 값을 찾아 현재 위치와 교환
     *   - 정렬 후 두 번째 요소를 반환
     * @param arr
     * @return
     */
    static int solution3(int[] arr) {
        selectionSort(arr);
        return arr[1];
    }

    static void selectionSort(int[] arr) {
        for (int i = 0; i < arr.length - 1; i++) {
            int minIndex = i;

            for (int j = i + 1; j < arr.length; j++) {
                if (arr[j] < arr[minIndex]) {
                    minIndex = j;
                }
            }

            swap(arr, i, minIndex);
        }
    }

    /**
     * 4. 삽입 정렬: 삽입 위치 찾기
     *   - 현재 값을 이미 정렬된 왼쪽 구간의 알맞은 위치에 삽입
     *   - 정렬 후 첫 번째 요소를 반환
     * @param arr
     * @return
     */
    static int solution4(int[] arr) {
        insertionSort(arr);
        return arr[0];
    }

    static void insertionSort(int[] arr) {
        for (int i = 1; i < arr.length; i++) {
            int current = arr[i];
            int j = i - 1;

            while (j >= 0 && arr[j] > current) {
                arr[j + 1] = arr[j];
                j--;
            }

            arr[j + 1] = current;
        }
    }

    /**
     * 5. 힙 정렬: 최대값 추출
     *   - 최대 힙을 만든 뒤 최대값을 한 번 제거
     *   - 다시 루트에 오는 값이 두 번째로 큰 값
     * @param arr
     * @return
     */
    static int solution5(int[] arr) {
        int heapSize = arr.length;

        buildMaxHeap(arr, heapSize);
        swap(arr, 0, heapSize - 1);
        heapSize--;
        heapify(arr, heapSize, 0);

        return arr[0];
    }

    static void buildMaxHeap(int[] arr, int heapSize) {
        for (int i = heapSize / 2 - 1; i >= 0; i--) {
            heapify(arr, heapSize, i);
        }
    }

    static void heapify(int[] arr, int heapSize, int rootIndex) {
        int largest = rootIndex;
        int left = rootIndex * 2 + 1;
        int right = rootIndex * 2 + 2;

        if (left < heapSize && arr[left] > arr[largest]) {
            largest = left;
        }

        if (right < heapSize && arr[right] > arr[largest]) {
            largest = right;
        }

        if (largest != rootIndex) {
            swap(arr, rootIndex, largest);
            heapify(arr, heapSize, largest);
        }
    }

    /**
     * 6. 병합 정렬: 분할 정복
     *   - 배열을 반으로 나누고 정렬된 두 구간을 병합
     *   - 길이가 짝수면 왼쪽 중간 값을 반환
     * @param arr
     * @return
     */
    static int solution6(int[] arr) {
        mergeSort(arr, 0, arr.length - 1);
        return arr[(arr.length - 1) / 2];
    }

    static void mergeSort(int[] arr, int left, int right) {
        if (left >= right) {
            return;
        }

        int mid = left + (right - left) / 2;
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);
        merge(arr, left, mid, right);
    }

    static void merge(int[] arr, int left, int mid, int right) {
        int[] sorted = new int[right - left + 1];
        int leftIndex = left;
        int rightIndex = mid + 1;
        int sortedIndex = 0;

        while (leftIndex <= mid && rightIndex <= right) {
            if (arr[leftIndex] <= arr[rightIndex]) {
                sorted[sortedIndex] = arr[leftIndex];
                leftIndex++;
            } else {
                sorted[sortedIndex] = arr[rightIndex];
                rightIndex++;
            }

            sortedIndex++;
        }

        while (leftIndex <= mid) {
            sorted[sortedIndex] = arr[leftIndex];
            leftIndex++;
            sortedIndex++;
        }

        while (rightIndex <= right) {
            sorted[sortedIndex] = arr[rightIndex];
            rightIndex++;
            sortedIndex++;
        }

        for (int i = 0; i < sorted.length; i++) {
            arr[left + i] = sorted[i];
        }
    }

    /**
     * 7. 퀵 정렬: 피벗 선택
     *   - 첫 번째 요소를 피벗으로 선택
     *   - 피벗보다 작은 값은 왼쪽, 큰 값은 오른쪽으로 나눈 뒤 재귀 정렬
     *   - 정렬 후 첫 번째 요소와 마지막 요소의 합을 반환
     *   - 근데 그렇다면.. 1 + 10 이라서 11이 맞는 것 같은데 문제 출력은 6이네 흠...
     * @param arr
     * @return
     */
    static int solution7(int[] arr) {
        quickSort(arr, 0, arr.length - 1);
        return arr[0] + arr[arr.length - 1];
    }

    static void quickSort(int[] arr, int left, int right) {
        if (left >= right) {
            return;
        }

        int pivotIndex = partition(arr, left, right);
        quickSort(arr, left, pivotIndex - 1);
        quickSort(arr, pivotIndex + 1, right);
    }

    static int partition(int[] arr, int left, int right) {
        int pivot = arr[left];
        int low = left + 1;
        int high = right;

        while (low <= high) {
            while (low <= right && arr[low] <= pivot) {
                low++;
            }

            while (high > left && arr[high] > pivot) {
                high--;
            }

            if (low < high) {
                swap(arr, low, high);
            }
        }

        swap(arr, left, high);
        return high;
    }

    static void swap(int[] arr, int firstIndex, int secondIndex) {
        int temp = arr[firstIndex];
        arr[firstIndex] = arr[secondIndex];
        arr[secondIndex] = temp;
    }
}
